peps/pep-0747.rst

```diff
@@ -0,0 +1,977 @@
+``TypeExpr[]`` is intended to be enforced only by static and runtime
+type checkers and need not be enforced by Python itself at runtime.
+
+
+.. _motivation:
+
+Motivation
+==========
+
+The introduction of ``TypeExpr`` allows new kinds of metaprogramming
+functions that operate on type expressions to be type-annotated and
+understood by type checkers.
+
+For example, here is a function that checks whether a value is
+assignable to a variable of a particular type, and if so returns the
+original value:
+
+::
+
+   def trycast[T](typx: TypeExpr[T], value: object) -> T | None: ...
+
+The use of ``TypeExpr[]`` and the type variable ``T`` enable the return
+type of this function to be influenced by a ``typx`` value passed at
+runtime, which is quite powerful.
+
+Here is another function that checks whether a value is assignable to a
+variable of a particular type, and if so returns ``True`` (as a special
+``TypeIs[]`` bool [#TypeIsPep]_):
+
+::
+
+   def isassignable[T](value: object, typx: TypeExpr[T]) -> TypeIs[T]: ...
+
+The use of ``TypeExpr[]`` and ``TypeIs[]`` together enables type
+checkers to narrow the return type appropriately depending on what type
+expression is passed in:
+
+::
+
+   request_json: object = ...
+   if isassignable(request_json, MyTypedDict):
+       assert_type(request_json, MyTypedDict)  # type is narrowed!
+
+That ``isassignable`` function enables a kind of enhanced ``isinstance``
+check which is useful for `checking whether a value decoded from JSON
+conforms to a particular structure`_ of nested ``TypedDict``\ s,
+``List``\ s, unions, ``Literal``\ s, and other types. This kind
+of check was alluded to in :pep:`PEP 589 <589#using-typeddict-types>` but could
+not be implemented at the time without a notation similar to
+``TypeExpr[]``.
+
+.. _checking whether a value decoded from JSON conforms to a particular structure: https://mail.python.org/archives/list/typing-sig@python.org/thread/I5ZOQICTJCENTCDPHLZR7NT42QJ43GP4/
+
+
+Why can’t ``type[]`` be used?
+-----------------------------
+
+One might think you could define the example functions above to take a
+``type[C]`` - which is syntax that already exists - rather than a
+``TypeExpr[T]``. However if you were to do that then certain type
+expressions like ``str | None`` - which are not class objects and
+therefore not ``type``\ s at runtime - would be rejected:
+
+::
+
+   # NOTE: Uses a type[C] parameter rather than a TypeExpr[T]
+   def trycast_type[C](typ: type[C], value: object) -> T | None: ...
+
+   trycast_type(str, 'hi')  # ok; str is a type
+   trycast_type(Optional[str], 'hi')  # ERROR; Optional[str] is not a type
+   trycast_type(str | int, 'hi')  # ERROR; (str | int) is not a type
+   trycast_type(MyTypedDict, dict(value='hi'))  # questionable; accepted by mypy 1.9.0
+
+To solve that problem, ``type[]`` could be widened to include the
+additional values allowed by ``TypeExpr``. However doing so would lose
```

Comment:
```suggestion
additional values allowed by ``TypeExpr``. However, doing so would lose
```

Reply:
The pause introduced by the added comma breaks the flow when I read the paragraph. Leaving out.